core/TypePtr.h

```diff
@@ -94,13 +107,13 @@ class TypePtr final {
         return static_cast<tagged_storage>(tag);
     }
 
-    static tagged_storage tagValue(Tag tag, uint32_t inlinedValue) {
+    static tagged_storage tagValue(Tag tag, uint64_t inlinedValue) {
+        // Ensure the upper bits aren't getting used for anything exciting.
+        ENFORCE_NO_TIMER((inlinedValue >> 48) == 0);
         auto val = tagToMask(tag);
 
-        // Store value into val.  It doesn't much matter where we put it in
-        // the upper 48 bits, but we put it in the uppermost 32 bits to
-        // ensure that retrieving it requires only a shift and no masking.
-        val |= static_cast<tagged_storage>(inlinedValue) << 32;
+        // Store value into val.  Put it in the same place as pointer values.
+        val |= static_cast<tagged_storage>(inlinedValue) << 16;
```

Comment:
Sorry I've been trying to piece together which bits correspond to what values in the `TypePtr`, depending on whether something is inlined or not.

I _think_ that since we're using `store & NOT_INLINED_MASK`, we need to shift the value to inline at least 17 bits?

I think i'd prefer keeping it at 32 bits just so that we can say "yep, we're only ever storing 32 bit values in here" rather than having to deal with non-word sizes like 47 bits.

Reply:
Anyways, I'm not sure if you find this useful or not, but for my own understanding of how things fit together, i put together these diagrams

```
TypePtr, !containsPtr():
  0x0000'0000'0000'0000
    ├───────┘    │ ├──┘
    │            │ └──▶ 16 bits for `TypePtr::Tag`
    │            └──▶ 1 bit for `NOT_INLINED_MASK`
    │                 Bit is 0, therefore inlined.
    └──▶ 32 bits for inlined value, from class
         created using TYPE_INLINED

TypePtr, containsPtr():
  0x0000'0000'0000'0000
    ├────────────┤ ├──┘
    │            │ └──▶ 16 bits for `TypePtr::Tag`
    │            └──▶ 1 bit for `NOT_INLINED_MASK`
    │                 Bit is 1, therefore not inlined.
    │                 Overlaps with LSB of pointer,
    │                 lowest 3 bits of pointer are
    │                 unused due to 8-byte alignment
    └──▶ 48 bits for pointer

TYPE(...) layout:
  0x0000'0000'0000'0000
    ├──┘ ├──┘ ├──···──┘
    │    │    └──▶ normal class fields class
    │    └──▶ C++ structure padding to align on 8 bytes
    └──▶ `Refcounted` (atomic counter via parent class)
```

if you'd like to include them in the source code, feel free. otherwise, i'll at least be able to refer back to this PR in case I ever want to see them in the future.

Reply:
This is a cool diagram!  Note that for `TYPE(...)`, the padding doesn't have to exist depending on the alignment of the initial fields in the class.

I'm not actually sure we need `NOT_INLINED_MASK` anymore, that might be a holdover from a previous `TypePtr` tagging scheme (it ought to be redundant with the tag value, unless we can have nullptr tagged pointers...which I don't think we can, but maybe?  jvilk had that concern about `ExpressionPtr`).  Will have to investigate.

I left shifted by 16 so that we could conveniently store "small" integers (32-bit unsigned values + 32-bit negative values) without thinking too much about things...oh, and so we can store `NameRef` + at least 1 bit for determining whether we're storing literal symbols or literal strings.

Reply:
> This is a cool diagram! Note that for TYPE(...), the padding doesn't have to exist depending on the alignment of the > initial fields in the class.

I realized that after posting and edited (i think maybe you responded to a pre-edit version of the comment). There's a caveat now.

> I'm not actually sure we need NOT_INLINED_MASK anymore, that might be a holdover from a previous TypePtr tagging scheme (it ought to be redundant with the tag value, unless we can have nullptr tagged pointers...which I don't think we can, but maybe? jvilk had that concern about ExpressionPtr). Will have to investigate.

sounds good, i'm happy to use the tag bits to determine whether the ptr is tagged or not

There is a comment on top of `NOT_INLINED_MASK` related to nullptr, fwiw

Reply:
Ah, the diagram is incorrect: we only use 8 bits for the tag (the entire low byte), and the second-lowest byte stores `NOT_INLINED_MASK`.  The remaining six bytes hold the actual value (inlined value or pointer, as appropriate), so there's no possible overlap between the value and `NOT_INLINED_MASK`...which is good, because otherwise that would mean our `ENFORCE` in `tagValue` wasn't firing on some very common cases!

Reply:
ah ok that makes sense, yeah there were multiple times when making that diagram that i thought one hex letter was one byte.

Reply:
(i have updated the chart)